tests/basic/pct.rs

```diff
@@ -94,3 +94,154 @@ fn yield_spin_loop_fair() {
 fn yield_spin_loop_unfair() {
     yield_spin_loop(false);
 }
+
+#[test]
+#[should_panic(expected = "called `Option::unwrap()` on a `None`")]
+// Based on Fig 1(a) of the PCT paper.  We model NULL pointer dereference with an Option unwrap
+fn figure1a_pct() {
+    const COUNT: usize = 5usize;
+    // n=2, d=1, so probability of finding the bug is at least 1/2
+    // So probability of hitting the bug in 20 iterations = 1 - (1 - 1/2)^20 > 99.99%
+    let scheduler = PctScheduler::new(1, 20);
+    let runner = Runner::new(scheduler, Default::default());
+    runner.run(|| {
+        let t1 = Arc::new(Mutex::new(None));
+        let t2 = Arc::clone(&t1);
+
+        thread::spawn(move || {
+            for _ in 0..COUNT {
+                thread::yield_now();
+            }
+            *t1.lock().unwrap() = Some(1);
+            for _ in 0..COUNT {
+                thread::yield_now();
+            }
+        });
+
+        thread::spawn(move || {
+            for _ in 0..COUNT {
+                thread::yield_now();
+            }
+            let _ = t2.lock().unwrap().unwrap();
+            for _ in 0..COUNT {
+                thread::yield_now();
+            }
+        });
+    });
+}
+
+// Based on Fig 1(b) from the PCT paper.  We model NULL pointer dereference with an Option unwrap.
+fn figure1b(num_threads: usize) {
+    assert!(num_threads >= 2);
+
+    let x1 = Arc::new(Mutex::new(Some(1)));
+    let x2 = Arc::clone(&x1);
+
+    // Optionally, spawn a bunch of threads that add scheduling choice points, each taking 5 steps
+    for _ in 0..num_threads - 2 {
+        thread::spawn(|| {
+            for _ in 0..5 {
+                thread::yield_now();
+            }
+        });
+    }
+
+    // Main worker threads take 10 steps each
+    thread::spawn(move || {
+        for _ in 0..5 {
+            thread::yield_now();
+        }
+        *x1.lock().unwrap() = None;
+        for _ in 0..4 {
+            thread::yield_now();
+        }
+    });
+
+    thread::spawn(move || {
+        for _ in 0..4 {
+            thread::yield_now();
+        }
+        let b = {
+            let b = x2.lock().unwrap().is_some();
+            b
+        };
+        if b {
+            let _ = x2.lock().unwrap().unwrap();
+        }
+        for _ in 0..4 {
+            thread::yield_now();
+        }
+    });
+}
+
+#[test]
+#[should_panic(expected = "called `Option::unwrap()` on a `None`")]
+fn figure1b_pct() {
+    // n=2, k=20, d=2, so probability of finding the bug in one iteration is at least 1/(2*20)
+    // So probability of hitting the bug in 200 iterations = 1 - (1 - 1/40)^200 > 99.36%
```

Comment:
My totally arbitrary target for these was 99.9%, maybe bump the iterations a little to get there (same for `figure_1c`).

Reply:
Done.